File: src/org/kde/kdeconnect/Plugins/SystemvolumePlugin/SystemVolumePlugin.java
/*
 * SPDX-FileCopyrightText: 2018 Nicolas Fella <nicolas.fella@gmx.de>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
*/

package org.kde.kdeconnect.Plugins.SystemvolumePlugin;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.kde.kdeconnect.NetworkPacket;
import org.kde.kdeconnect.Plugins.Plugin;
import org.kde.kdeconnect.Plugins.PluginFactory;
import org.kde.kdeconnect_tp.R;

import java.util.ArrayList;
import java.util.Collection;
import java.util.concurrent.ConcurrentHashMap;

@PluginFactory.LoadablePlugin
public class SystemVolumePlugin extends Plugin {

    private final static String PACKET_TYPE_SYSTEMVOLUME = "kdeconnect.systemvolume";
    private final static String PACKET_TYPE_SYSTEMVOLUME_REQUEST = "kdeconnect.systemvolume.request";

    public interface SinkListener {
        void sinksChanged();
    }

    private final ConcurrentHashMap<String, Sink> sinks;
    private final ArrayList<SinkListener> listeners;

    public SystemVolumePlugin() {
        sinks = new ConcurrentHashMap<>();
        listeners = new ArrayList<>();
    }

    @Override
    public String getDisplayName() {
        return context.getResources().getString(R.string.pref_plugin_systemvolume);
    }

    @Override
    public String getDescription() {
        return context.getResources().getString(R.string.pref_plugin_systemvolume_desc);
    }

    @Override
    public boolean onPacketReceived(NetworkPacket np) {

        if (np.has("sinkList")) {
            sinks.clear();

            try {
                JSONArray sinkArray = np.getJSONArray("sinkList");

                for (int i = 0; i < sinkArray.length(); i++) {
                    JSONObject sinkObj = sinkArray.getJSONObject(i);
                    Sink sink = new Sink(sinkObj);
                    sinks.put(sink.getName(), sink);
                }
            } catch (JSONException e) {
                Log.e("KDEConnect", "Exception", e);
            }

            for (SinkListener l : listeners) {
                l.sinksChanged();
            }

        } else {
            String name = np.getString("name");
            if (sinks.containsKey(name)) {
                if (np.has("volume")) {
                    sinks.get(name).setVolume(np.getInt("volume"));
                }
                if (np.has("muted")) {
                    sinks.get(name).setMute(np.getBoolean("muted"));
                }
            }
        }
        return true;
    }

    void sendVolume(String name, int volume) {
        NetworkPacket np = new NetworkPacket(PACKET_TYPE_SYSTEMVOLUME_REQUEST);
        np.set("volume", volume);
        np.set("name", name);
        device.sendPacket(np);
    }

    void sendMute(String name, boolean mute) {
        NetworkPacket np = new NetworkPacket(PACKET_TYPE_SYSTEMVOLUME_REQUEST);
        np.set("muted", mute);
        np.set("name", name);
        device.sendPacket(np);
    }

    void requestSinkList() {
        NetworkPacket np = new NetworkPacket(PACKET_TYPE_SYSTEMVOLUME_REQUEST);
        np.set("requestSinks", true);
        device.sendPacket(np);
    }

    @Override
    public boolean hasMainActivity() {
        return false;
    }

    @Override
    public boolean displayInContextMenu() {
        return false;
    }

    @Override
    public String[] getSupportedPacketTypes() {
        return new String[]{PACKET_TYPE_SYSTEMVOLUME};
    }

    @Override
    public String[] getOutgoingPacketTypes() {
        return new String[]{PACKET_TYPE_SYSTEMVOLUME_REQUEST};
    }

    Collection<Sink> getSinks() {
        return sinks.values();
    }

    void addSinkListener(SinkListener listener) {
        listeners.add(listener);
    }

    void removeSinkListener(SinkListener listener) {
        listeners.remove(listener);
    }

}

File: src/main/java/org/kde/kdeconnect/plugins/systemvolume/SystemVolumePlugin.kt
/*
 * SPDX-FileCopyrightText: 2018 Nicolas Fella <nicolas.fella@gmx.de>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
*/
package org.kde.kdeconnect.plugins.systemvolume

import android.util.Log
import androidx.annotation.VisibleForTesting
import org.json.JSONException
import org.kde.kdeconnect.NetworkPacket
import org.kde.kdeconnect.plugins.Plugin
import org.kde.kdeconnect.plugins.PluginFactory.LoadablePlugin
import org.kde.kdeconnect_tp.R
import java.util.concurrent.ConcurrentHashMap

@LoadablePlugin
class SystemVolumePlugin : Plugin() {
    interface SinkListener {
        fun sinksChanged()
    }

    private val sinkMap: ConcurrentHashMap<String, Sink> = ConcurrentHashMap()
    private val listeners: MutableList<SinkListener> = mutableListOf()

    override val displayName: String
        get() = context.resources.getString(R.string.pref_plugin_systemvolume)

    override val description: String
        get() = context.resources.getString(R.string.pref_plugin_systemvolume_desc)

    override fun onPacketReceived(np: NetworkPacket): Boolean {
        if ("sinkList" in np) {
            sinkMap.clear()

            try {
                val sinkArray = checkNotNull(np.getJSONArray("sinkList"))
                for (i in 0..< sinkArray.length()) {
                    val sinkObj = sinkArray.getJSONObject(i)
                    val sink = Sink(sinkObj)
                    sinkMap[sink.name] = sink
                }
            } catch (e: JSONException) {
                Log.e("KDEConnect", "Exception", e)
            }

            synchronized(listeners) {
                for (l in listeners) {
                    l.sinksChanged()
                }
            }
        } else {
            val name = np.getString("name")
            val sink = sinkMap[name]
            if (sink != null) {
                if ("volume" in np) {
                    sink.setVolume(np.getInt("volume"))
                }
                if ("muted" in np) {
                    sink.setMute(np.getBoolean("muted"))
                }
                if ("enabled" in np) {
                    sink.isDefault = np.getBoolean("enabled")
                }
            }
        }
        return true
    }

    internal fun sendVolume(name: String, volume: Int) {
        val np = NetworkPacket(PACKET_TYPE_SYSTEMVOLUME_REQUEST)
        np["volume"] = volume
        np["name"] = name
        device.sendPacket(np)
    }

    internal fun sendMute(name: String, mute: Boolean) {
        val np = NetworkPacket(PACKET_TYPE_SYSTEMVOLUME_REQUEST)
        np["muted"] = mute
        np["name"] = name
        device.sendPacket(np)
    }

    internal fun sendEnable(name: String) {
        val np = NetworkPacket(PACKET_TYPE_SYSTEMVOLUME_REQUEST)
        np["enabled"] = true
        np["name"] = name
        device.sendPacket(np)
    }

    override val supportedPacketTypes: Array<String> = arrayOf(PACKET_TYPE_SYSTEMVOLUME)
    override val outgoingPacketTypes: Array<String> = arrayOf(PACKET_TYPE_SYSTEMVOLUME_REQUEST)

    val sinks: MutableCollection<Sink>
        get() = sinkMap.values

    internal fun addSinkListener(listener: SinkListener) {
        synchronized(listeners) {
            listeners.add(listener)
        }
    }

    internal fun removeSinkListener(listener: SinkListener) {
        synchronized(listeners) {
            listeners.remove(listener)
        }
    }

    companion object {
        private const val PACKET_TYPE_SYSTEMVOLUME = "kdeconnect.systemvolume"
        private const val PACKET_TYPE_SYSTEMVOLUME_REQUEST = "kdeconnect.systemvolume.request"
    }
}
